cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/ShardedHBaseQueueStrategy.java

```diff
@@ -111,11 +124,22 @@ public QueueScanner createScanner(ConsumerConfig consumerConfig,
     int caching = (int) (1.1 * numRows / distributorBuckets);
     shardedScan.setCaching(caching);
 
-    shardedScan.setStartRow(getShardedKey(consumerConfig, consumerConfig.getInstanceId(), scan.getStartRow()));
-    shardedScan.setStopRow(getShardedKey(consumerConfig, consumerConfig.getInstanceId(), scan.getStopRow()));
+    if (scan.getStartRow().length > 0) {
+      byte[] rowKey = getShardedKey(consumerConfig, consumerConfig.getInstanceId(), scan.getStartRow());
+      if (useActualRowKeys) {
+        rowKey = rowKeyDistributor.getDistributedKey(rowKey);
+      }
+      shardedScan.setStartRow(rowKey);
+    }
 
-    ResultScanner scanner = DistributedScanner.create(hTable, shardedScan, rowKeyDistributor, scansExecutor);
-    return new HBaseQueueScanner(scanner, numRows, ROW_KEY_CONVERTER);
+    if (scan.getStopRow().length > 0) {
+      byte[] rowKey = getShardedKey(consumerConfig, consumerConfig.getInstanceId(), scan.getStopRow());
+      if (useActualRowKeys) {
+        rowKey = rowKeyDistributor.getDistributedKey(rowKey);
```

Comment:
How does it work now if you set the Scan start/stop row to the actual HBase rowKey (meaning included the salted byte)? Isn't that the DistributedScanner.create() expects unsalted rowKey?